### ISSUE TITLE
Inconsistent return type when filtering a list

### ISSUE THREAD

**Issue by obi1kenobi** [orientechnologies/orientdb#7811]
## Expected behavior  
`SELECT out().name FROM V` and `SELECT out()[@this INSTANCEOF 'V'].name FROM V` should return the same output, with equal types for each returned row.

## Actual behavior  
`SELECT out().name FROM V` is consistent -- the returned data is always a list.

However, `SELECT out()[@this INSTANCEOF 'V'].name FROM V` is not consistent even though it should be an equivalent query to the above. When only one element is to be returned, instead of returning a list of one element, the query unwraps the list and simply returns the element itself.

## Steps to reproduce
Setup:  
```
CREATE VERTEX V SET name = 'foo'
CREATE VERTEX V SET name = 'bar'
CREATE EDGE E FROM (SELECT * FROM V WHERE name = 'foo') TO (SELECT * FROM V WHERE name = 'bar')
```

`SELECT out().name FROM V` returns two rows:
```
["bar"]
[]
```

`SELECT out()[@this INSTANCEOF 'V'].name FROM V` also returns two rows:
```
"bar"
[]
```

This is very confusing and unexpected. The type of the returned column is sometimes a list, and sometimes a string, which unnecessarily complicates the handling code.

Would it be possible to make filtering on the list always return a list, even if of one element?

This is currently causing a bug in the [GraphQL compiler](https://github.com/kensho-technologies/graphql-compiler) project, when someone uses a `@fold` directive and applies filters -- the calculated return type is a list type, but OrientDB sometimes returns a non-list.

**Comment by luigidellaquila** [comment]
Hi @obi1kenobi 

Yes, it's a known behavior (it's wrong IMHO, but for backward compatibility I cannot change it in v 2.2.x). 
As a work-around, you can use `list()` function to always transform the result to a list, eg.

```
SELECT list(out()[@this INSTANCEOF 'V'].name) FROM V
```

Good news is that this behavior is already fixed and made consistent in v 3.0, see http://orientdb.com/docs/3.0.x/sql/SQL-Syntax.html#square-bracket-filters

I hope it helps

Thanks

Luigi

**Comment by obi1kenobi** [comment]
Hi @luigidellaquila, thanks for the prompt reply. Unfortunately, that workaround doesn't produce equivalent behavior.

`SELECT list(out()[@this INSTANCEOF 'V'].name) FROM V` returns only a single row:
```
["bar"]
```
rather than the two rows `SELECT out().name FROM V` returns:
```
["bar"]
[]
```

I think this is because the `list()` operator tries to aggregate across all rows. For example, if we add an edge from `bar` to `foo` like so `CREATE EDGE E FROM (SELECT * FROM V WHERE name = 'bar') TO (SELECT * FROM V WHERE name = 'foo')`, then `SELECT out().name FROM V` outputs two rows:
```
["bar"]
["foo"]
```
but `SELECT list(out()[@this INSTANCEOF 'V'].name) FROM V` outputs only a single row:
```
["bar", "foo"]
```

I understand that you cannot change the square bracket filtering behavior for backwards compatibility. Would it be possible to just add a working workaround that does the right thing?

Thanks!

**Comment by luigidellaquila** [comment]
Oh, right... let me check...

**Comment by luigidellaquila** [comment]
Hi @obi1kenobi 

You have two alternatives:

- add a second null parameter, so that the list() function acts as an in-line function instead of an aggregate one
```
SELECT list(out()[@this INSTANCEOF 'V'].name, null) FROM V
```

- use `.asList()` method
```
SELECT out()[@this INSTANCEOF 'V'].name.asList() FROM V
```

Just consider that there is a little difference between the two: the first one returns an empty list when no results match the square bracket filtering, while the second one just returns a null value

I hope it helps

Thanks

Luigi

**Comment by obi1kenobi** [comment]
I think those work. Just to clarify though, it seems like the second one still returns an empty list. I queried with `SELECT out()[@this INSTANCEOF 'V'].name.asList() FROM V` and got:
```
["bar"]
[]
```

I then confirmed with `SELECT out()[@this INSTANCEOF 'V'].name.asList().size() FROM V` and got:
```
1
0
```

I think this is correct behavior, I'd find it unexpected if `.asList()` returned `null`, but I wanted to verify this is intentional and not going to change in the future :)

Thanks again!
